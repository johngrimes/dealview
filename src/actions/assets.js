// @flow

import type { Asset, AssetWithId, AssetMap } from 'types/assets/asset'

type PutAssetRequestAction = {
  type: 'PUT_ASSET_REQUEST',
  asset: Asset,
}
type PutAssetSuccessAction = {
  type: 'PUT_ASSET_SUCCESS',
  asset: AssetWithId
}
type PutAssetFailureAction = {
  type: 'PUT_ASSET_FAILURE',
  error: string|null
}

type DeleteAssetRequestAction = {
  type: 'DELETE_ASSET_REQUEST',
  id: string,
}
type DeleteAssetSuccessAction = {
  type: 'DELETE_ASSET_SUCCESS',
  id: string
}
type DeleteAssetFailureAction = {
  type: 'DELETE_ASSET_FAILURE',
  error: string|null
}

type LoadAssetsRequestAction = {
  type: 'LOAD_ASSETS_REQUEST'
}
type LoadAssetsSuccessAction = {
  type: 'LOAD_ASSETS_SUCCESS',
  assets: AssetMap
}
type LoadAssetsFailureAction = {
  type: 'LOAD_ASSETS_FAILURE',
  error: string|null
}

export type AssetAction = PutAssetRequestAction
                        | PutAssetSuccessAction
                        | PutAssetFailureAction
                        | DeleteAssetRequestAction
                        | DeleteAssetRequestAction
                        | DeleteAssetFailureAction
                        | LoadAssetsRequestAction
                        | LoadAssetsSuccessAction
                        | LoadAssetsFailureAction

export const putAssetRequest = (asset: Asset): PutAssetRequestAction => {
  return {
    type: 'PUT_ASSET_REQUEST',
    asset,
  }
}

export const putAssetSuccess = (asset: AssetWithId): PutAssetSuccessAction => {
  return {
    type: 'PUT_ASSET_SUCCESS',
    asset,
  }
}

export const putAssetFailure = (error: string|null): PutAssetFailureAction => {
  return {
    type: 'PUT_ASSET_FAILURE',
    error,
  }
}

export const deleteAssetRequest = (id: string): DeleteAssetRequestAction => {
  return {
    type: 'DELETE_ASSET_REQUEST',
    id,
  }
}

export const deleteAssetSuccess = (id: string): DeleteAssetSuccessAction => {
  return {
    type: 'DELETE_ASSET_SUCCESS',
    id,
  }
}

export const deleteAssetFailure = (error: string|null): DeleteAssetFailureAction => {
  return {
    type: 'DELETE_ASSET_FAILURE',
    error,
  }
}

export const loadAssetsRequest = (): LoadAssetsRequestAction => {
  return { type: 'LOAD_ASSETS_REQUEST' }
}

export const loadAssetsSuccess = (assets: AssetMap): LoadAssetsSuccessAction => {
  return {
    type: 'LOAD_ASSETS_SUCCESS',
    assets,
  }
}

export const loadAssetsFailure = (error: string|null): LoadAssetsFailureAction => {
  return {
    type: 'LOAD_ASSETS_FAILURE',
    error,
  }
}
