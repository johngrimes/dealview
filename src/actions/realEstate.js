// @flow

import type { RealEstate, RealEstateWithId, RealEstateMap } from 'types/assets/realEstate'

type PutRealEstateRequestAction = {
  type: 'PUT_REAL_ESTATE_REQUEST',
  realEstate: RealEstate,
}
type PutRealEstateSuccessAction = {
  type: 'PUT_REAL_ESTATE_SUCCESS',
  realEstate: RealEstateWithId,
}
type PutRealEstateFailureAction = {
  type: 'PUT_REAL_ESTATE_FAILURE',
  error: string|null,
}

type DeleteRealEstateRequestAction = {
  type: 'DELETE_REAL_ESTATE_REQUEST',
  id: string,
}
type DeleteRealEstateSuccessAction = {
  type: 'DELETE_REAL_ESTATE_SUCCESS',
  id: string,
}
type DeleteRealEstateFailureAction = {
  type: 'DELETE_REAL_ESTATE_FAILURE',
  error: string|null,
}

type LoadRealEstateRequestAction = {
  type: 'LOAD_REAL_ESTATE_REQUEST'
}
type LoadRealEstateSuccessAction = {
  type: 'LOAD_REAL_ESTATE_SUCCESS',
  realEstate: RealEstateMap,
}
type LoadRealEstateFailureAction = {
  type: 'LOAD_REAL_ESTATE_FAILURE',
  error: string|null,
}

export type RealEstateAction = PutRealEstateRequestAction
                             | PutRealEstateSuccessAction
                             | PutRealEstateFailureAction
                             | DeleteRealEstateRequestAction
                             | DeleteRealEstateRequestAction
                             | DeleteRealEstateFailureAction
                             | LoadRealEstateRequestAction
                             | LoadRealEstateSuccessAction
                             | LoadRealEstateFailureAction

export const putRealEstateRequest = (realEstate: RealEstate): PutRealEstateRequestAction => {
  return {
    type: 'PUT_REAL_ESTATE_REQUEST',
    realEstate,
  }
}

export const putRealEstateSuccess = (realEstate: RealEstateWithId): PutRealEstateSuccessAction => {
  return {
    type: 'PUT_REAL_ESTATE_SUCCESS',
    realEstate,
  }
}

export const putRealEstateFailure = (error: string|null): PutRealEstateFailureAction => {
  return {
    type: 'PUT_REAL_ESTATE_FAILURE',
    error,
  }
}

export const deleteRealEstateRequest = (id: string): DeleteRealEstateRequestAction => {
  return {
    type: 'DELETE_REAL_ESTATE_REQUEST',
    id,
  }
}

export const deleteRealEstateSuccess = (id: string): DeleteRealEstateSuccessAction => {
  return {
    type: 'DELETE_REAL_ESTATE_SUCCESS',
    id,
  }
}

export const deleteRealEstateFailure = (error: string|null): DeleteRealEstateFailureAction => {
  return {
    type: 'DELETE_REAL_ESTATE_FAILURE',
    error,
  }
}

export const loadRealEstateRequest = (): LoadRealEstateRequestAction => {
  return { type: 'LOAD_REAL_ESTATE_REQUEST' }
}

export const loadRealEstateSuccess = (realEstate: RealEstateMap): LoadRealEstateSuccessAction => {
  return {
    type: 'LOAD_REAL_ESTATE_SUCCESS',
    realEstate,
  }
}

export const loadRealEstateFailure = (error: string|null): LoadRealEstateFailureAction => {
  return {
    type: 'LOAD_REAL_ESTATE_FAILURE',
    error,
  }
}
