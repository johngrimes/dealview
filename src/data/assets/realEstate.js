// @flow

import moment from 'moment'

import { getDatabase, save as saveToDatabase, load as loadFromDatabase, loadAll as loadAllFromDatabase } from '../database.js'
import { AddressEmpty } from '../commonTypes.js'
import { ValuationsEmpty } from '../../components/forms/ValuationsInput.js'
import type { Values } from '../database.js'
import type { AddressValues } from '../commonTypes.js'
import type { Valuation, Valuations } from '../../components/forms/ValuationsInput.js'
import type { AssetValues } from './asset.js'

export type RealEstateValues = {
  id: string,
  name: string,
  address: AddressValues,
  notes: string,
  valuations: Valuations
}

export const RealEstateEmpty = {
  id: '',
  name: '',
  address: AddressEmpty,
  notes: '',
  valuations: ValuationsEmpty
}

export const save = (values: Values): Promise<string> => {
  return saveToDatabase('Asset.RealEstate', values)
}

export const load = (id: string): Promise<RealEstateValues> => {
  return loadFromDatabase('Asset.RealEstate', id)
}

export const loadAll = (): Promise<RealEstateValues[]> => {
  return loadAllFromDatabase('Asset.RealEstate')
}

export const assetSubscriber = (eventType: string, content: RealEstateValues): void => {
  getAssetRecord(content.id).then(existingAsset => {
    const lastValuation = getLastValuation(content)
    const asset: AssetValues = {
      type: 'RealEstate',
      instanceId: content.id,
      name: content.name
    }
    if (lastValuation) { asset.lastValuation = lastValuation.amount }
    if (existingAsset) { asset.id = existingAsset.id }
    saveToDatabase('Asset', asset).catch(error => { console.error(error) })
  })
}

export const getLastValuation = (realEstate: RealEstateValues): Valuation|false => {
  if (realEstate.valuations.length === 0) {
    return false
  }
  const sortedValuations = realEstate.valuations.slice().sort(compareValuationsByDate)
  const idxFuture = sortedValuations.findIndex(v => { moment(v.date, 'YYYY-MM-DD').valueOf() > moment().valueOf() })
  const slicedValuations = sortedValuations.slice(0, idxFuture)
  return slicedValuations[slicedValuations.length - 1]
}

const compareValuationsByDate = (a: Valuation, b: Valuation): number => {
  const [milliA, milliB] = [a, b].map(v => { return v.date ? moment(v.date).valueOf() : 0 })
  return milliA - milliB
}

const getAssetRecord = (realEstateId: string): Promise<AssetValues|false> => {
  return new Promise((resolve, reject) => {
    getDatabase().then(db => {
      const transaction = db.transaction(['Asset'], 'readonly')
      const store = transaction.objectStore('Asset')
      const index = store.index('type, instanceId')
      const request = index.get(['RealEstate', realEstateId])
      request.onsuccess = event => { resolve(event.target.result) }
      request.onerror = () => { resolve(false) }
    })
  })
}
