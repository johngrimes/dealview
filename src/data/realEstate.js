import { AddressDefaults } from './commonTypes.js'
import { saleNegFilter } from './valuations.js'

export const RealEstateDefaults = {
  name: '',
  address: AddressDefaults,
  notes: '',
  purchaseDate: '',
  valuations: [],
}

export const realEstateToAsset = realEstate => {
  const asset = {}
  asset.type = 'RealEstate'
  asset.name = realEstate.name
  asset.startDate = realEstate.purchaseDate
  asset.endDate = realEstate.saleDate
  // Sale valuations are omitted from the valuations that get stored against the asset.
  // The asset was never worth that much on our balance sheet - that value immediately went to cash, or whatever else.
  asset.valuations = realEstate.valuations.filter(saleNegFilter)
  if (typeof realEstate.id === 'string') {
    asset.id = realEstate.id
  }
  return asset
}
