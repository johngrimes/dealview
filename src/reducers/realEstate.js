// @flow

import { createObjectsReducer, InitialObjectStoreState } from 'reducers/objects'
import type { ObjectStoreState } from 'reducers/objects'
import type { RealEstateMap } from 'types/assets/realEstate'
import type { RealEstateAction } from 'actions/realEstate'

export type RealEstateState = ObjectStoreState & { +objects: RealEstateMap }

export const InitialRealEstateState: RealEstateState = InitialObjectStoreState

const RealEstateReducer: (RealEstateState, RealEstateAction) => RealEstateState =
  createObjectsReducer('REAL_ESTATE', 'REAL_ESTATE')

export default RealEstateReducer
