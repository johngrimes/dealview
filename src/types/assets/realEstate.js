// @flow

import { AddressDefaults } from 'types/commonTypes'
import type { Address } from 'types/commonTypes'
import type { Asset } from 'types/assets/asset'
import type { Valuations } from 'components/forms/ValuationsInput'

export type RealEstate = {
  name: string,
  address: Address,
  notes: string,
  purchaseDate: string,
  saleDate?: string,
  valuations: Valuations
}

export type RealEstateWithId = RealEstate & { id: string }

export type RealEstateMap = { [id: string]: RealEstate }

export const RealEstateDefaults = {
  name: '',
  address: AddressDefaults,
  notes: '',
  purchaseDate: '',
  valuations: [],
}

export const realEstateToAsset = (realEstate: RealEstate): Asset => {
  const asset = {}
  asset.type = 'RealEstate'
  asset.name = realEstate.name
  asset.startDate = realEstate.purchaseDate
  asset.endDate = realEstate.saleDate
  asset.valuations = realEstate.valuations
  if (typeof realEstate.id === 'number') { asset.instanceId = realEstate.id.toString() }
  return asset
}
