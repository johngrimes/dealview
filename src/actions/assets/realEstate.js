// @flow

import { putObject, deleteObject, getAllObjects } from 'db/db'
import AssetActions from 'actions/assets/assets'
import { realEstateToAsset } from 'types/assets/realEstate'
import { createObjectActions } from 'actions/objects'
import type { RealEstate } from 'types/assets/realEstate'
import type { Thunk } from 'types/commonTypes'

const objectStore = 'Asset.RealEstate'

const actions = createObjectActions('REAL_ESTATE', 'REAL_ESTATE',
                                    'RealEstate', 'RealEstate')

const putRealEstate = (realEstate: RealEstate): Thunk => {
  return dispatch => {
    dispatch(actions.putRealEstateRequest(realEstate))
    return new Promise(async (resolve, reject) => {
      try {
        const saved = await putObject(objectStore, realEstate)
        dispatch(actions.putRealEstateSuccess(saved))
        await dispatch(AssetActions.putAsset(realEstateToAsset(saved)))
        resolve(saved)
      } catch (error) {
        dispatch(actions.putRealEstateFailure(error))
        reject(error)
      }
    })
  }
}

const deleteRealEstate = (id: string): Thunk => {
  return dispatch => {
    dispatch(actions.deleteRealEstateRequest(id))
    return new Promise(async (resolve, reject) => {
      try {
        const deletedId = await deleteObject(objectStore, id)
        dispatch(AssetActions.deleteAsset(deletedId))
        dispatch(actions.deleteRealEstateSuccess(deletedId))
        resolve(deletedId)
      } catch (error) {
        dispatch(actions.deleteRealEstateFailure(error))
        reject(error)
      }
    })
  }
}

const loadRealEstate = (): Thunk => {
  return dispatch => {
    dispatch(actions.loadRealEstateRequest())
    return new Promise((resolve, reject) => {
      getAllObjects(objectStore).then(assets => {
        dispatch(actions.loadRealEstateSuccess(assets))
        resolve(assets)
      }).catch(error => {
        dispatch(actions.loadRealEstateFailure(error))
        reject(error)
      })
    })
  }
}

export default {
  ...actions,
  putRealEstate,
  deleteRealEstate,
  loadRealEstate,
}
