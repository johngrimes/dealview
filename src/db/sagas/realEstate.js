// @flow

import { takeEvery, call, put, all } from 'redux-saga/effects'

import * as RealEstateActions from 'actions/realEstate'
import { putObject, deleteObject, getAllObjects } from 'db/db'
import type { PutRealEstateRequestAction, DeleteRealEstateRequestAction } from 'actions/realEstate'

const objectStore = 'Asset.RealEstate'

export function* watchPutRealEstate() {
  yield takeEvery('PUT_REAL_ESTATE_REQUEST', putRealEstate)
}

export function* putRealEstate(action: PutRealEstateRequestAction) {
  try {
    const saved = yield call(putObject, objectStore, action.realEstate)
    yield put(RealEstateActions.putRealEstateSuccess(saved))
  } catch (error) {
    yield put(RealEstateActions.putRealEstateFailure(error))
  }
}

function* watchDeleteRealEstate() {
  yield takeEvery('DELETE_REAL_ESTATE_REQUEST', deleteRealEstate)
}

function* deleteRealEstate(action: DeleteRealEstateRequestAction) {
  try {
    const id = yield call(deleteObject, objectStore, action.id)
    yield put(RealEstateActions.deleteRealEstateSuccess(id))
  } catch (error) {
    yield put(RealEstateActions.deleteRealEstateFailure(error))
  }
}

function* watchLoadRealEstate() {
  yield takeEvery('LOAD_REAL_ESTATE_REQUEST', loadRealEstate)
}

function* loadRealEstate() {
  try {
    const realEstate = yield call(getAllObjects, objectStore)
    yield put(RealEstateActions.loadRealEstateSuccess(realEstate))
  } catch (error) {
    yield put(RealEstateActions.loadRealEstateFailure(error))
  }
}

export default function* watchRealEstate(): void {
  yield all([
    watchPutRealEstate(),
    watchDeleteRealEstate(),
    watchLoadRealEstate(),
  ])
}
