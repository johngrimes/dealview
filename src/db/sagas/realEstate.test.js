/* global expect */

import expectGen from 'expect-gen'

import watchRealEstate from 'db/sagas/realEstate'
import { putRealEstateRequest } from 'actions/realEstate'
import { validRealEstate1 } from 'fixtures/realEstate'

describe('putRealEstate', () => {
  it('should dispatch correct actions when successful', () => {
    const effects = expectGen(watchRealEstate)
      .yields(putRealEstateRequest(validRealEstate1))
      .finishes()
      .toJSON()
    expect(effects).toMatchSnapshot()
  })
})
