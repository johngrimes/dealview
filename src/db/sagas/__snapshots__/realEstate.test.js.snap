// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`putRealEstate should dispatch correct actions when successful 1`] = `
Array [
  Object {
    "done": false,
    "value": Object {
      "@@redux-saga/IO": true,
      "ALL": Array [
        Object {},
        Object {},
        Object {},
      ],
    },
  },
  Object {
    "done": true,
  },
]
`;
